File: testsuite/mcpserver.py
#!/usr/bin/python3
"""
Test Utility: MCP Server Communication Harness

Description:
    This module provides the 'McpServer' helper class used across the test suite
    to spin up, communicate with, and manage the life cycle of the snapper MCP server
    subprocess. It implements the JSON-RPC 2.0 communication protocol over standard I/O (stdio).

Key Features:
    - Subprocess Management: Spawns the MCP server binary (default: '/usr/bin/mcp-server-snapper')
      or uses the location specified by the 'MCPSERVER' environment variable.
    - Handshake Protocol: Automatically executes the MCP initialization handshake protocol
      ('initialize' request with appropriate client capabilities and info) upon instantiation.
    - JSON-RPC Communication: Encapsulates sending requests, handling ID incrementation,
      and parsing incoming JSON-RPC responses.
"""

import os
import subprocess
import json
import sys


class McpServer:

    def __init__(self):

        self.request_id = 0

        mcp_server = os.environ.get('MCPSERVER', "/usr/bin/mcp-server-snapper")

        self.process = subprocess.Popen([ mcp_server ], stdin = subprocess.PIPE,
                                        stdout = subprocess.PIPE, stderr = sys.stderr,
                                        universal_newlines = True)

        print("--- Sending Initialize ---")

        self.send_request("initialize", {
            "protocolVersion": "2025-11-25",
            "capabilities": {},
            "clientInfo": { "name": "testsuite", "version": "1.0.0" }
        })


    def send_request(self, method, params = None):

        self.request_id += 1

        request = {
            "jsonrpc": "2.0",
            "id": self.request_id,
            "method": method,
            "params": params or {}
        }

        self.process.stdin.write(json.dumps(request) + "\n")
        self.process.stdin.flush()

        response = self.process.stdout.readline()
        if not response:
            raise Exception("Failed to read response.")

        json_response = json.loads(response)

        print(json.dumps(json_response, indent = 2))

        return json_response

File: testsuite/mcpserver/mcpserver.go
package mcpserver

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"strings"
)

type McpServer struct {
	cmd    *exec.Cmd
	stdin  io.WriteCloser
	stdout io.ReadCloser
	reader *bufio.Reader
}

func NewMcpServer() (*McpServer, error) {
	mcpServerBin := os.Getenv("MCPSERVER")
	if mcpServerBin == "" {
		mcpServerBin = "/usr/bin/mcp-server-snapper"
	}

	cmd := exec.Command(mcpServerBin)
	cmd.Stderr = os.Stderr

	stdin, err := cmd.StdinPipe()
	if err != nil {
		return nil, fmt.Errorf("failed to create stdin pipe: %w", err)
	}

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, fmt.Errorf("failed to create stdout pipe: %w", err)
	}

	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("failed to start MCP server: %w", err)
	}

	srv := &McpServer{
		cmd:    cmd,
		stdin:  stdin,
		stdout: stdout,
		reader: bufio.NewReader(stdout),
	}

	fmt.Println("--- Sending Initialize ---")
	resp, err := srv.SendRequest("initialize", map[string]any{
		"protocolVersion": "2025-11-25",
		"capabilities":    map[string]any{},
		"clientInfo": map[string]string{
			"name":    "testsuite",
			"version": "1.0.0",
		},
	})
	if err != nil {
		srv.Close()
		return nil, fmt.Errorf("initialize failed: %w", err)
	}

	// Verify protocol version
	result, ok := resp["result"].(map[string]any)
	if !ok {
		srv.Close()
		return nil, fmt.Errorf("malformed initialize response result")
	}
	if result["protocolVersion"] != "2025-11-25" {
		srv.Close()
		return nil, fmt.Errorf("unexpected protocol version: %v", result["protocolVersion"])
	}

	return srv, nil
}

func (s *McpServer) Close() {
	if s.stdin != nil {
		s.stdin.Close()
	}
	if s.cmd != nil {
		_ = s.cmd.Process.Kill()
		_ = s.cmd.Wait()
	}
}

func (s *McpServer) SendRequest(method string, params any) (map[string]any, error) {
	req := map[string]any{
		"jsonrpc": "2.0",
		"id":      1,
		"method":  method,
		"params":  params,
	}

	bytes, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	if _, err := s.stdin.Write(append(bytes, '\n')); err != nil {
		return nil, fmt.Errorf("failed to write request: %w", err)
	}

	line, err := s.reader.ReadString('\n')
	if err != nil {
		return nil, fmt.Errorf("failed to read response line: %w", err)
	}

	line = strings.TrimSpace(line)
	if line == "" {
		return nil, fmt.Errorf("empty response received")
	}

	var resp map[string]any
	if err := json.Unmarshal([]byte(line), &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	// Print response pretty-printed
	pretty, _ := json.MarshalIndent(resp, "", "  ")
	fmt.Println(string(pretty))

	return resp, nil
}
